### PR TITLE
Add spegel distributed registry mirror

### PR DESCRIPTION
**Pull request by brandond**
#### Proposed Changes ####

This embeds [spegel](https://github.com/XenitAB/spegel), a distributed registry mirror, into the K3s supervisor. In addition to reducing traffic against upstream registries, this also allows airgap images preloaded onto one node to be transparently shared to other nodes as needed.

spegel consists of a registry API backed by the local containerd image store, and a [distributed hash table](https://pkg.go.dev/github.com/libp2p/go-libp2p-kad-dht) that allows nodes to gossip about which images and blobs they have available.

The local registry mirror is injected into the containerd config as the first mirror endpoint, followed by the user-selected endpoints, and finally the default endpoint. If any node in the cluster has an image, it will be pulled from that node, instead of from the registry or registry mirror.

The embedded registry is enabled at a cluster level via a CLI flag. When enabled, all nodes will listen on port 5001 for P2P traffic, secured by a preshared key. Agents will have a new listener on port 6443 that hosts the registry endpoint. The registry API is served over HTTPS, and requires a valid client certificate for access.

This appears to add ~4.4MB to the size of the K3s release artifact.

Checklist:
* [x] Write up ADR supporting this new feature and get team+PM approval
* [x] Get spegel working embedded in supervisor endpoint
* [x] Add agents to spegel cluster, instead of just proxying back to servers
* [x] Remove hardcoded values and wire into node config
* [x] Add CLI flag to enable, instead of always being enabled
* [x] Require TLS auth for registry endpoint
* [x] Require private network with preshared key
* [x] Add tests.
* [x] Add ipv6 support and push upstream. 
  Handled by upstream in https://github.com/XenitAB/spegel/pull/284
* [ ] Push embedding, https, and nonblocking containerd client support upstream  
  https://github.com/XenitAB/spegel/compare/main...k3s-io:spegel:k3s-main

Builds on:
* [x] https://github.com/k3s-io/k3s/pull/8973

Inspired by:
* https://github.com/k3s-io/k3s/pull/8554
* https://github.com/XenitAB/spegel/pull/281

Resolves:
* https://github.com/k3s-io/k3s/issues/8781
* https://github.com/k3s-io/k3s/issues/9026

**Use:**
1. Start servers with `--embedded-registry`
2. In order to enable distributed mirroring of an upstream registry, reference it in the mirror section of `registries.yaml` on every node that you want to participate in the sharing of images. The registry does not need to have any endpoints, although it may. For example, this is a valid configuration:  
    ```yaml
    mirrors:
      docker.io:
      registry.k8s.io:
      gcr.io:
      quay.io:
      ghcr.io:
    ```
3. Nodes also may be started with `--disable-default-registry-endpoint`, in which case images will **only** be available via airgap tarball, distributed mirror, or user-configured mirror (in order of precendece)
4. Check metrics:  
  `kubectl get --raw /api/v1/nodes/NODENAME/proxy/metrics | grep -F 'spegel_`

#### Types of Changes ####

new feature

#### Verification ####

See above

#### Testing ####

E2e test added.

#### Linked Issues ####

* TBD

#### User-Facing Change ####

```release-note

```

### CODE REVIEW

**Comment by phillebaba** [comment]
We can check off ipv6 support as I have merged https://github.com/XenitAB/spegel/pull/284.

**Comment by harsimranmaan** [comment]
Thanks Brandon. I am really happy to see such good progress on this one. I can help test this. If you need support convincing the team or the PMs, you can count my vote (and perhaps @adrianmoye's  too) :)

File: pkg/cli/cmds/server.go
Comment (harsimranmaan) [inline pkg/cli/cmds/server.go:495]: When you say embedded containterd, do you mean k3s managed containerd or is there a separate option for using embedded containerd?

**Review by buroa — APPROVED** [review]

**Comment by onedr0p** [comment]
It's truly amazing the work you put into adding this @brandond and so quickly, can't wait to try it out. ❤️

File: pkg/spegel/spegel.go
Comment (harsimranmaan) [inline pkg/spegel/spegel.go:97]: Do you think that these values can be made configurable by using an override from say the `registries.yaml` in the future

```yaml
...
embedded-registry:
  router-port: 5001
```

Comment (brandond) [reply pkg/spegel/spegel.go:97]: No, I don't think we should try to extend registries.yaml to contain field specific to spegel.

I don't really want to add a CLI flag for this either, but if I do it will be set server-side and passed down to clients, as I think it should be the same cluster-wide. Perhaps I would use an env var for it - `LIBP2P_PORT` or something.

Comment (brandond) [reply pkg/spegel/spegel.go]: OK, I added an env var. I don't want to **ENCOURAGE** people to change it, but it is possible.

**Comment by twistedgrim** [comment]
I just learned about Spegel and tried out it out. Freaking amazing work! Looking forward to this!

File: pkg/spegel/spegel.go
Comment (harsimranmaan) [reply pkg/spegel/spegel.go]: Thanks Brandon. Introducing a new port can be hard as there are chances of conflict with another service. Making it configurable would be helpful in my case as we have moved from gravity to k3s and would like to reuse a port instead of asking customers to open another one.

**Comment by rlex** [comment]
spegel exposes some useful metrics, will they be exposed in case of running embedded spegel? Cannot find anything related in PR

**Comment by onedr0p** [comment]
It would be very helpful useful to have those metrics exposed, their grafana dashboard is quite nice for reviewing what's going on.

**Comment by brandond** [comment]
I hadn't wired that up yet but I think it's possible. Were you looking at the libp2p metrics, or image metrics, or both?

**Comment by brandond** [comment]
@onedr0p @rlex metrics for both spegel and libp2p should now be exposed on the same endpoint as the rest of the K3s Kubernetes metrics

**Comment by onedr0p** [comment]
Perfect, that should work great!

**Comment by dereknola** [comment]
Looking through the code, this seems GTM, however, I would like for at least a basic E2E test (expanding on the existing e2e startup test would be easiest) with this feature enabled to be written before we merge this PR.

**Comment by harsimranmaan** [comment]
Looks like upstream support for exposing the spegel api is the only outstanding item. @phillebaba maybe you can help with that. Wish you all a prosperous 2024.

**Comment by brandond** [comment]
I think we'll probably merge this as is; if there are additional changes to spegel to allow for improvement to embedded support we can drop our fork.

**Comment by harsimranmaan** [comment]
Cannot wait to get my hands on it.

**Review by dereknola — DISMISSED** [review]

**Review by harsimranmaan — APPROVED** [review]

File: pkg/cli/cmds/agent.go
Comment (manuelbuil) [inline pkg/cli/cmds/agent.go:224]: Why would a user want to activate this? Isn't it good to have a fallback?

File: tests/e2e/embeddedmirror/embeddedmirror_test.go
Comment (manuelbuil) [inline tests/e2e/embeddedmirror/embeddedmirror_test.go:86]: In other e2e tests we are passing the flag `--kubeconfig=" + kubeConfigFile` and you are not. Perhaps it is not needed since it is k3s but just wanted to warn you. Did you verify that the test passes successfully?

File: pkg/spegel/spegel.go
Comment (manuelbuil) [inline pkg/spegel/spegel.go:55]: Do you foresee a case where the node has the p2p node-address annotation filled but the enabled label is false? Just wondering if we need the annotation and the label

File: tests/e2e/embeddedmirror/embeddedmirror_test.go
Comment (manuelbuil) [inline tests/e2e/embeddedmirror/embeddedmirror_test.go:90]: What about adding the `"p2p." + version.Program + ".cattle.io/enabled"` to verify that the label is present with `true`?

**Review by manuelbuil — COMMENTED** [review]
Great Brad! It is a big PR, hard to review in detail but it looks very good in general. I just have a few questions/comments.

IIUC, to enable this feature, the user would need to:
* Create /etc/rancher/k3s/registries.yaml with at least one mirror
* embedded-registry: true in the servers
Right?

EDIT: The answer is in the description of the PR! Sorry

File: pkg/cli/cmds/agent.go
Comment (adrianmoye) [reply pkg/cli/cmds/agent.go:224]: As a user... I really like this flag if I understand what it does correctly.

This would make my life easier when air-gapping workloads, making testing that I have mirrored all of the required containers easier.

From a security perspective, I don't like my production non air-gapped clusters pulling untrusted containers down from external registries. They should only be running known trusted images, from trusted local sources.

I would use this flag on all production systems, and for some testing/development work.

Comment (brandond) [reply pkg/cli/cmds/agent.go:224]: We've had folks open issues about it in the past, expressing surprise that containerd was falling back to the default public endpoint despite having a local mirror configured. Now that we have an option to disable this on the containerd side, I figured it'd be nice to expose that here in k3s.

File: tests/e2e/embeddedmirror/embeddedmirror_test.go
Comment (brandond) [reply tests/e2e/embeddedmirror/embeddedmirror_test.go:86]: I just copied this from the existing `privateregistry` e2e tests, so whatever works there should work here.

File: pkg/spegel/spegel.go
Comment (brandond) [reply pkg/spegel/spegel.go:55]: The label is used for matching when listing nodes using a label selector. You can't match/filter nodes based on annotations, and I wanted to be able to select only nodes that have this enabled.

When I started tinkering with this, the embedded registry was a node-level option instead of cluster-level, and I left the node selector logic here in case we do in the future want to allow some nodes to opt out of using the embedded registry.

File: tests/e2e/embeddedmirror/embeddedmirror_test.go
Comment (brandond) [reply tests/e2e/embeddedmirror/embeddedmirror_test.go:90]: That shouldn't be necessary; I'm not trying to validate that the label is present - just that the agent nodes (which  have the default registry endpoint disabled) are able to pull the deployment image via the distributed registry from the servers, which are able to pull from Docker Hub.

Comment (manuelbuil) [reply tests/e2e/embeddedmirror/embeddedmirror_test.go:90]: I was suggesting it as a way to confirm that the distributed registry is working. I was wondering if there could be a faulty case where the cluster deploys correctly but the distributed registry is not working well and then images are pulled correctly too but coming from upstream. I guess activating the `default registry endpoint disabled` option makes that faulty case impossible

**Review by manuelbuil — APPROVED** [review]

File: tests/e2e/embeddedmirror/embeddedmirror_test.go
Comment (brandond) [reply tests/e2e/embeddedmirror/embeddedmirror_test.go:90]: Maybe I'll add a check of the metrics to confirm that it came from the cache.

Comment (brandond) [reply tests/e2e/embeddedmirror/embeddedmirror_test.go:90]: Added additional tests.

File: docs/adrs/embedded-registry.md
Comment (brandond) [reply docs/adrs/embedded-registry.md:32]: It is used both locally by containerd, and by spegel on other nodes when providing content for containerd. Spegel docs cover the actual flow from containerd -> spegel proxy -> spegel image store -> containerd image store.

**Review by manuelbuil — APPROVED** [review]

**Review by galal-hussein — APPROVED** [review]

File: pkg/agent/containerd/containerd.go
Comment (galal-hussein) [inline pkg/agent/containerd/containerd.go:274]: sorry small comment here, is there a chance that image.Labels may be nil, in that case we can expect a panic here

Comment (brandond) [reply pkg/agent/containerd/containerd.go:274]: No. You can check keys in a nil map without panicing. You just can't assign to a nil map.

**Pull request merged**
